docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Pod initialization
+
+@smarterclayton
+
+March 2016
+
+## Proposal and Motivation
+
+Within a pod there is a need to initialize local data or adapt to the current
+cluster environment that is not easily achieved in the current container model.
+Containers start in parallel after volumes are mounted, leaving no opportunity
+for coordination between containers without specialization of the image. If
+two containers need to share common initialization data, both images must
+be altered to cooperate using filesystem or network semantics, which introduces
+coupling between images. Likewise, if an image requires configuration in order
+to start and that configuration is environment dependent, the image must be
+altered to add the necessary templating or retrieval.
+
+This proposal introduces the concept of an **init container**, one or more
```

Comment:
Thinking about this - hooks are (today) running in a container.  In
OpenShift, we have deployment level hooks that create a pod.  We could
imagine hooks that make TCP or HTTP requests in other contexts.
PreStartContainers?  I think "PreStart' is more consistent with existing
API terminology than "Init".  On the other hand, would we add "PostStop"
containers?  Is "PreStart" container hooks the same as a pre-start pod hook?


On Mon, Apr 18, 2016 at 6:12 PM, Vish Kannan <notifications@github.com>
wrote:

> In docs/proposals/container-init.md
> <https://github.com/kubernetes/kubernetes/pull/23666#discussion_r60143604>
> :
>
> > +
> > +March 2016
> > +
> > +## Proposal and Motivation
> > +
> > +Within a pod there is a need to initialize local data or adapt to the current
> > +cluster environment that is not easily achieved in the current container model.
> > +Containers start in parallel after volumes are mounted, leaving no opportunity
> > +for coordination between containers without specialization of the image. If
> > +two containers need to share common initialization data, both images must
> > +be altered to cooperate using filesystem or network semantics, which introduces
> > +coupling between images. Likewise, if an image requires configuration in order
> > +to start and that configuration is environment dependent, the image must be
> > +altered to add the necessary templating or retrieval.
> > +
> > +This proposal introduces the concept of an **init container**, one or more
>
> @smarterclayton <https://github.com/smarterclayton> +1 for your
> rationale. What is your thought on renaming "init" containers to be pod
> level "pre-start" hooks?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23666/files/7efdc3d36197f1ca452339d49abbc87163fb0e75#r60143604>
>